### ISSUE TITLE
ShouldBeEquivalentTo fails when comparing objects from the System namespace

### ISSUE THREAD

**Issue by ConradLang** [fluentassertions/fluentassertions#305]
Hi

I'm using the expected object pattern to assert that the exception thrown is equivalent to the expected exception.
It appears that ShouldBeEquivalentTo works for objects outside the `System` namespace but not for objects in the `System` namespace.
This behaviour feels inconsistent - is it a bug?

Here are three examples which can be used to reproduce the error.

**1. The following snippet shows an example class defined in the `System` namespace and a `NotSystem` namespace and a couple of tests to show the unexpected failure when the class is defined in the system namespace.**

I've created classes in the `System` namespace to show that it appears the issue is specific to when the class is in the `System` namespace.

``` c#
namespace System
{
    public class Something
    {
        public string TestValue => "Something in System namespace";
    }
}

namespace NotSystem
{
    public class Something
    {
        public string TestValue => "Something not in System namespace";
    }
}

namespace FluentValidationTests
{
    [TestClass]
    public class UnitTest
    {
        [TestMethod]
        public void ShouldBeEquivalentTo_With_Object_From_System_Namespace_Should_Pass()
        {
            var systemObject1 = new System.Something();
            var systemObject2 = new System.Something();

            // This fails because the "Something" class is in the "System" namespace.
            systemObject2.ShouldBeEquivalentTo(systemObject1);
        }

        [TestMethod]
        public void ShouldBeEquivalentTo_With_Object_From_Non_System_Namespace_Should_Pass()
        {
            var systemObject1 = new NotSystem.Something();
            var systemObject2 = new NotSystem.Something();

            // This passes because the "Something" class is _not_ in the "System" namespace.
            systemObject2.ShouldBeEquivalentTo(systemObject1);
        }
    }
}

```

_The `ShouldBeEquivalentTo_With_Object_From_System_Namespace_Should_Pass` test fails with:_

```
Test Name:  ShouldBeEquivalentTo_With_Object_From_System_Namespace_Should_Pass
Test FullName:  FluentValidationTests.UnitTest.ShouldBeEquivalentTo_With_Object_From_System_Namespace_Should_Pass
Test Source:    C:\Projects\TestFluentValidation\FluentValidationTests\UnitTest.cs : line 105
Test Outcome:   Failed
Test Duration:  0:00:00.0147143

Result StackTrace:  
at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message) in C:\projects\fluentassertions-vf06b\Src\Shared\Execution\LateBoundTestFramework.cs:line 26
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message) in C:\projects\fluentassertions-vf06b\Src\FluentAssertions.Net40\Execution\TestFrameworkProvider.cs:line 42
   at FluentAssertions.Execution.CollectingAssertionStrategy.ThrowIfAny(IDictionary`2 context) in C:\projects\fluentassertions-vf06b\Src\Core\Execution\CollectingAssertionStrategy.cs:line 50
   at FluentAssertions.Execution.AssertionScope.Dispose() in C:\projects\fluentassertions-vf06b\Src\Core\Execution\AssertionScope.cs:line 252
   at FluentAssertions.Equivalency.EquivalencyValidator.AssertEquality(EquivalencyValidationContext context) in C:\projects\fluentassertions-vf06b\Src\Core\Equivalency\EquivalencyValidator.cs:line 37
   at FluentAssertions.AssertionExtensions.ShouldBeEquivalentTo[T](T subject, Object expectation, Func`2 config, String because, Object[] reasonArgs) in C:\projects\fluentassertions-vf06b\Src\Shared\AssertionExtensions.cs:line 610
   at FluentAssertions.AssertionExtensions.ShouldBeEquivalentTo[T](T subject, Object expectation, String because, Object[] reasonArgs) in C:\projects\fluentassertions-vf06b\Src\Shared\AssertionExtensions.cs:line 572
   at FluentValidationTests.UnitTest.ShouldBeEquivalentTo_With_Object_From_System_Namespace_Should_Pass() in C:\Projects\TestFluentValidation\FluentValidationTests\UnitTest.cs:line 109
Result Message: 
Expected subject to be 

System.Something
{
   TestValue = "Something in System namespace"
}, but found 

System.Something
{
   TestValue = "Something in System namespace"
}.

With configuration:
- Use declared types and members
- Compare enums by value
- Match member by name (or throw)
- Be strict about the order of items in byte arrays

```

**2. The following snippet shows an example custom exception class defined in the `System` namespace and a `NotSystem` namespace and a couple of tests to show the unexpected failure when the class is defined in the system namespace.**

``` c#
namespace System
{
    public class CustomException : System.Exception
    {
        public CustomException(string message) : base(message) { }
    }
}

namespace NotSystem
{
    public class CustomException : System.Exception
    {
        public CustomException(string message) : base(message) { }
    }
}

namespace FluentValidationTests
{
    [TestClass]
    public class UnitTest
    {
        [TestMethod]
        public void ShouldBeEquivalentTo_With_Custom_Exception_In_System_Namespace_Should_Pass()
        {
            var systemObject1 = new System.CustomException("A custom exception");
            System.CustomException systemObject2;
            try
            {
                throw new System.CustomException("A custom exception");
            }
            catch (System.CustomException ex)
            {
                systemObject2 = ex;
            }

            // Exclude non-public properties and properties that depend on where the exception is thrown.
            var propertiesToExclude = typeof (System.CustomException).GetProperties(BindingFlags.Instance | BindingFlags.NonPublic).Select(p => p.Name).ToList();
            propertiesToExclude.Add(nameof(System.Exception.Source));
            propertiesToExclude.Add(nameof(System.Exception.StackTrace));
            propertiesToExclude.Add(nameof(System.Exception.TargetSite));

            // This fails because the "CustomException" class is in the "System" namespace.
            systemObject2.ShouldBeEquivalentTo(
                systemObject1,
                options => options.Excluding(ex => propertiesToExclude.Contains(ex.SelectedMemberInfo.Name)));
        }

        [TestMethod]
        public void ShouldBeEquivalentTo_With_Custom_Exception_In_Non_System_Namespace_Should_Pass()
        {
            var systemObject1 = new NotSystem.CustomException("A custom exception");
            NotSystem.CustomException systemObject2;
            try
            {
                throw new NotSystem.CustomException("A custom exception");
            }
            catch (NotSystem.CustomException ex)
            {
                systemObject2 = ex;
            }

            // Exclude non-public properties and properties that depend on where the exception is thrown.
            var propertiesToExclude = typeof (NotSystem.CustomException).GetProperties(BindingFlags.Instance | BindingFlags.NonPublic).Select(p => p.Name).ToList();
            propertiesToExclude.Add(nameof(System.Exception.Source));
            propertiesToExclude.Add(nameof(System.Exception.StackTrace));
            propertiesToExclude.Add(nameof(System.Exception.TargetSite));

            // This passes because the "CustomException" class is _not_ in the "System" namespace.
            systemObject2.ShouldBeEquivalentTo(
                systemObject1,
                options => options.Excluding(ex => propertiesToExclude.Contains(ex.SelectedMemberInfo.Name)));
        }
    }
}
```

_The `ShouldBeEquivalentTo_With_Custom_Exception_In_System_Namespace_Should_Pass` test fails with:_

```
Test Name:  ShouldBeEquivalentTo_With_Custom_Exception_In_System_Namespace_Should_Pass
Test FullName:  FluentValidationTests.UnitTest.ShouldBeEquivalentTo_With_Custom_Exception_In_System_Namespace_Should_Pass
Test Source:    C:\Projects\TestFluentValidation\FluentValidationTests\UnitTest.cs : line 57
Test Outcome:   Failed
Test Duration:  0:00:00.0062068

Result StackTrace:  
at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message) in C:\projects\fluentassertions-vf06b\Src\Shared\Execution\LateBoundTestFramework.cs:line 26
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message) in C:\projects\fluentassertions-vf06b\Src\FluentAssertions.Net40\Execution\TestFrameworkProvider.cs:line 42
   at FluentAssertions.Execution.CollectingAssertionStrategy.ThrowIfAny(IDictionary`2 context) in C:\projects\fluentassertions-vf06b\Src\Core\Execution\CollectingAssertionStrategy.cs:line 50
   at FluentAssertions.Execution.AssertionScope.Dispose() in C:\projects\fluentassertions-vf06b\Src\Core\Execution\AssertionScope.cs:line 252
   at FluentAssertions.Equivalency.EquivalencyValidator.AssertEquality(EquivalencyValidationContext context) in C:\projects\fluentassertions-vf06b\Src\Core\Equivalency\EquivalencyValidator.cs:line 37
   at FluentAssertions.AssertionExtensions.ShouldBeEquivalentTo[T](T subject, Object expectation, Func`2 config, String because, Object[] reasonArgs) in C:\projects\fluentassertions-vf06b\Src\Shared\AssertionExtensions.cs:line 610
   at FluentValidationTests.UnitTest.ShouldBeEquivalentTo_With_Custom_Exception_In_System_Namespace_Should_Pass() in C:\Projects\TestFluentValidation\FluentValidationTests\UnitTest.cs:line 74
Result Message: 
Expected subject to be System.CustomException with message "A custom exception"
, but found System.CustomException with message "A custom exception"
     at FluentValidationTests.UnitTest.ShouldBeEquivalentTo_With_Custom_Exception_In_System_Namespace_Should_Pass() in C:\Projects\TestFluentValidation\FluentValidationTests\UnitTest.cs:line 62
.

With configuration:
- Use declared types and members
- Compare enums by value
- Exclude member when value(FluentValidationTests.UnitTest+<>c__DisplayClass1_0).propertiesToExclude.Contains(ex.SelectedMemberInfo.Name)
- Match member by name (or throw)
- Be strict about the order of items in byte arrays
```

**3. The following snippet shows the behaviour when a `System.Exception` is compared.**

This shows that the .NET `System.Exception` (and any other system exception) is affected.

``` c#
namespace FluentValidationTests
{
    [TestClass]
    public class UnitTest
    {
        [TestMethod]
        public void ShouldBeEquivalentTo_With_System_Exception_Should_Pass()
        {
             var propertiesToExclude = typeof(System.Exception).GetProperties(BindingFlags.Instance | BindingFlags.NonPublic).Select(p => p.Name).ToList();
            propertiesToExclude.Add(nameof(System.Exception.Source));
            propertiesToExclude.Add(nameof(System.Exception.StackTrace));
            propertiesToExclude.Add(nameof(System.Exception.TargetSite));

            var systemObject1 = new System.Exception("Just a System.Exception");
            var systemObject2 = new System.Exception("Just a System.Exception");

            systemObject2.ShouldBeEquivalentTo(
                systemObject1,
                options => options.Excluding(ex => propertiesToExclude.Contains(ex.SelectedMemberInfo.Name)));
       }
    }
}
```

_The test fails with:_

```
Test Name:  ShouldBeEquivalentTo_Should_Pass_With_System_Exceptions_But_Fails
Test FullName:  FluentValidationTests.UnitTest.ShouldBeEquivalentTo_Should_Pass_With_System_Exceptions_But_Fails
Test Source:    C:\Projects\TestFluentValidation\FluentValidationTests\UnitTest.cs : line 41
Test Outcome:   Failed
Test Duration:  0:00:00.3572282

Result StackTrace:  
at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message) in C:\projects\fluentassertions-vf06b\Src\Shared\Execution\LateBoundTestFramework.cs:line 26
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message) in C:\projects\fluentassertions-vf06b\Src\FluentAssertions.Net40\Execution\TestFrameworkProvider.cs:line 42
   at FluentAssertions.Execution.CollectingAssertionStrategy.ThrowIfAny(IDictionary`2 context) in C:\projects\fluentassertions-vf06b\Src\Core\Execution\CollectingAssertionStrategy.cs:line 50
   at FluentAssertions.Execution.AssertionScope.Dispose() in C:\projects\fluentassertions-vf06b\Src\Core\Execution\AssertionScope.cs:line 252
   at FluentAssertions.Equivalency.EquivalencyValidator.AssertEquality(EquivalencyValidationContext context) in C:\projects\fluentassertions-vf06b\Src\Core\Equivalency\EquivalencyValidator.cs:line 37
   at FluentAssertions.AssertionExtensions.ShouldBeEquivalentTo[T](T subject, Object expectation, Func`2 config, String because, Object[] reasonArgs) in C:\projects\fluentassertions-vf06b\Src\Shared\AssertionExtensions.cs:line 610
   at FluentValidationTests.UnitTest.ShouldBeEquivalentTo_Should_Pass_With_System_Exceptions_But_Fails() in C:\Projects\TestFluentValidation\FluentValidationTests\UnitTest.cs:line 50
Result Message: 
Expected subject to be System.Exception with message "Just a System.Exception"
, but found System.Exception with message "Just a System.Exception"
.

With configuration:
- Use declared types and members
- Compare enums by value
- Exclude member when value(FluentValidationTests.UnitTest+<>c__DisplayClass0_0).propertiesToExclude.Contains(ex.SelectedMemberInfo.Name)
- Match member by name (or throw)
- Be strict about the order of items in byte arrays
```

**Comment by dennisdoomen** [comment]
This behavior is by-design. Anything that's in the same namespace as the standard .NET privimites is treated as value types that we shoudn't inspect for properties. This behavior can be changed globally through [this line of code](https://github.com/dennisdoomen/fluentassertions/blob/develop/Src/Core/AssertionOptions.cs#L35).

**Comment by ConradLang** [comment]
Thanks for the reply and workaround. I've changed the `IsValueType` predicate in my test initialize so that `System.Exception` and its derivatives are not recognised as value types.
_This is not meant to support defining classes in the `System` namespace since you would not normally do that. I did so just to demonstrate the behaviour._

```c#
AssertionOptions.IsValueType = type => (type.Namespace == typeof(int).Namespace) && !(type == typeof(System.Exception) || type.IsSubclassOf(typeof(System.Exception)));
```

This is fine for altering the global behaviour but would it be useful to have something like `EquivalencyAssertionOptions.ComparingByReference` to allow overriding the behaviour locally instead of globally?

**Comment by dennisdoomen** [comment]
Yeah, that would make sense indeed.

**Comment by dennisdoomen** [comment]
While looking at this, I realized there's already a local `ComparingByValue` method on the options.

**Comment by ConradLang** [comment]
The `ComparingByValue` method adds the type to the list of value types ([SelfReferenceEquivalencyAssertionOptions Line 480](https://github.com/dennisdoomen/fluentassertions/blob/develop/Src/Core/Equivalency/SelfReferenceEquivalencyAssertionOptions.cs#L480)) which then is used to determine if the subject is a value type by checking against the list of value types and then the global options.IsValueType delegate  ([SelfReferenceEquivalencyAssertionOptions Line 182](https://github.com/dennisdoomen/fluentassertions/blob/develop/Src/Core/Equivalency/SelfReferenceEquivalencyAssertionOptions.cs#L182)).

The suggested `ComparingByReference` method could add to a "referenceTypes" list that is then specifically excluded e.g.
```c#
bool IEquivalencyAssertionOptions.IsValueType(Type type)
{
    return !referenceTypes.Contains(type) && (valueTypes.Contains(type) || AssertionOptions.IsValueType(type));
}
```

**Comment by TimMurphy** [comment]
@dennisdoomen why is this by design? I too am comparing actual exception object to expected.

**Comment by dennisdoomen** [comment]
It was kind of a shortcut to prevent FA from recursively traversing the properties of primitive types.

**Comment by dennisdoomen** [comment]
This has been fixed in Fluent Assertions 5.0, where we changed the algorithm on how value types are detected. See also [the docs](https://fluentassertions.com/objectgraphs/#value-types).
